### PR TITLE
Using OpmLog instead of explicit logger

### PR DESCRIPTION
**Pull request by joakim-hove**
Followup to: https://github.com/OPM/opm-parser/pull/398

### CODE REVIEW

File: examples/sim_fibo_ad.cpp [hunk 1]
```diff
@@ -104,29 +107,48 @@ try
...
+    }
+
+    std::string logFile = output_dir + "/LOGFILE.txt";
+    Opm::ParserPtr parser(new Opm::Parser());
+    {
+        std::shared_ptr<Opm::StreamLog> streamLog = std::make_shared<Opm::StreamLog>(logFile , Opm::Log::DefaultMessageTypes);
+        std::shared_ptr<Opm::CounterLog> counterLog = std::make_shared<Opm::CounterLog>(Opm::Log::DefaultMessageTypes);
+
+        Opm::OpmLog::addBackend( "STREAM" , streamLog );
+        Opm::OpmLog::addBackend( "COUNTER" , streamLog );
```
Comment (atgeirr) [inline examples/sim_fibo_ad.cpp]: Looks like a bug, what is the counter log by the way?

File: examples/sim_fibo_ad.cpp [hunk 2]
```diff
@@ -104,29 +107,48 @@ try
...
-        }
-        std::cerr << "error while parsing the deck file: " << e.what() << "\n";
+        std::cerr << "Failed to create valid ECLIPSESTATE object. See logfile: " << logFile << std::endl;
+        std::cerr << "Exception caught: " << e.what() << std::endl;
         return EXIT_FAILURE;
     }
 
-    if (logger->size() > 0) {
-        std::cerr << "Issues found while parsing the deck file:\n";
-        logger->printAll(std::cerr);
```
Comment (atgeirr) [inline examples/sim_fibo_ad.cpp]: A question: how is this functionality handled after this PR?

**Comment by joakim-hove** [comment]
Thank you for looking into this; I will unfortunately have quite long turnaround time today.

>Looks like a bug,

That was certainly a bug - I will push an update.

> what is the counter log by the way?

`CounterLog` is the remains of @andlaus original `ParserLog` implementation; an important part of the original implementation was to keep track of the all warnings, errors and so on, so that you could ask the logger at the end: "OK buddy - how many warnings?". That functionality is now in the `CounterLog` backend: https://github.com/joakim-hove/opm-parser/blob/singleton-example/opm/parser/eclipse/OpmLog/CounterLog.cpp implementation.

>A question: how is this functionality handled after this PR?

This functionality is not 100% reproduced, the current state is:

1. Iff you have supplied a `StreamLog` backend to the logger the actual log messages will be there.
2. Iff you have supplied a `CounterLog` backend you can fetch that from the logger and ask it:

```C++
auto counterLog = OpmLog::getBackend<CounterLog>("COUNTER");
size_t numWarning = counterLog->getNummessages(Log::MessageType::Warning)
```

I'm grateful for input - both the on the actual implementation, and how it can/should be used. It would be a grate shame if the developers in the end said: "Ahhh - that logging shait; I'll just use std::cout".

**Comment by atgeirr** [comment]
Thanks for the info. What is the default behaviour with this: will there be warnings printed as they are encountered or will they reside in the log silently until asked for (and they are currently not asked for it seems)?

About usage: I'll imagine we'll have more to say about that when we actually start using it, but from the PR in opm-parser I can already see that I would like an addWarning() convenience function (and addNote() etc... a complete set of them would be nice). More readable and lowers the barrier to usage I think.

**Comment by joakim-hove** [comment]
>What is the default behaviour with this: will there be warnings printed as they are encountered or will they reside in the log silently until asked for (and they are currently not asked for it seems)?

As long as you have not added any logging backends the call `OpmLog::addMessage()` will essentially be a no-op, i.e. this https://github.com/joakim-hove/opm-parser/blob/singleton-example/opm/parser/eclipse/OpmLog/OpmLog.cpp#L36 will return immediately.

When you have added one or multiple backends the behaviour will be accordingly; in the current autodiff example PR the `StreamLog` instance will collect all messages in the file "LOGFILE.txt", the `CounterLog` instance will count errors/warnings/... but currently not do anything about it.

>I can already see that I would like an addWarning() convenience function (and addNote()  ...)

Yes I understand; but before adding those convenience functions it would be good to have an agreement[1] on which types to have, because I guess the moment those convenience functions are added the general `addMessage()` method will only be for those specially interested.

My very loose suggestions:

    Exception: An exception is thrown.
    Error: A serious error - probably not recoverable.
    Warning / Problem: Eclipse has both - don't know the distinction; but we can find out.
    Info: Innocent information of various kinds.
    Debug: All sorts of drivel?

 
 

[1]: Including a policy for when to use which message types.

**Comment by joakim-hove** [comment]
OK - I have updated the list of default messageTypes: https://github.com/joakim-hove/opm-parser/blob/singleton-example/opm/parser/eclipse/OpmLog/LogUtil.hpp#L28 - somewhat inspired by ECLIPSE.

**Comment by joakim-hove** [comment]
OK - unless someone objects I will merge this code tomorrow afternoon.

**Comment by atgeirr** [comment]
I will comment on the messageTypes:
 - They seem intended as independent flags (being 1, 2, 4, 8 etc.) rather than an ascending chain where each higher level includes the lower ones. As I see it, if I log an Error, I should not have to say that, "btw, this is also a Debug, Info and Warning message", that is implicit. What is the use case then, for the individual flags?
 - I think that Debug, Info and Warning are fine, but I have a little more trouble with the intended relationship between Error, Problem and Bug. If Problem is something that we will often continue from anyway (such as a convergence failure which is your example) it should be lower in the chain than Error. Alternatively, we could split it according to fatality: Error and FatalError (or just Fatal). I do not see how Bug should be useful at this point. I think just having Error for all fatal things (and use Warning for non-fatal things) would be the simplest though.
 - What is the intended relationship with error propagation? Mostly we use exceptions for this, but there are some places where we use error return-values. I do not think that adding an Error log message should trigger an exception, but perhaps it should be the other way around? For example OPM_THROW should be remade to use addError() instead of printing a message.

**Comment by joakim-hove** [comment]
OK - thank you again for comments:

> They seem intended as independent flags (being 1, 2, 4, 8 etc.) rather than an ascending chain 

Yes - that is true; and I think it should be like that. I think that you are maybe mixing the implementation of a logging system, and the end user configuration of the logging. That end user configuration will in this case be translated to which messagetypes the various backends accepts.

> As I see it, if I log an Error, I should not have to say that, "btw, this is also a Debug, Info and Warning message", that is implicit. What is the use case then, for the individual flags?

Here I disagree - if I log an Error it is an Error - not a warning? Maybe I misunderstand you.

>I think that Debug, Info and Warning are fine, but I have a little more trouble with the intended relationship between Error, Problem and Bug

I have no strong feelings for these values, the suggestion was *loosely* inspired by the `MESSAGES` keyword of ECLIPSE. I am more than happy to change it in various directions.

> [...] it should be lower in the chain than Error

I think that statement illustrates the problem of a hardcoded strictly ordered hierarchy of message types - why do have to order them in the implementation.

We must agree on the different types of messages and when they should be emitted, and then the end user application (i.e. `sim_fibo_ad` in our case) can configure which messages it will handle, and where they such go. When the application/end user configures logging I agree that a hierarchy can make sense, but that does not have to (and in my opinion should not) carry all the way down to the implementation of logging.

> What is the intended relationship with error propagation? 

Well - first of all we must agree on this. My *suggestion* would be that OPM_THROW added a
```
OpmLog::addError("Fatal ..."); // or addBug / addException ....
```

With this code:
```C++
std::shared_pointer<StreamLog> logFile = std::make_shared("LOGFILE.TXT" , Log::DefaultMessaegTypes);
std::shared:pointer<StreamLog> cerrLog = std::make_shared(std::cerr , Log::MessaegType::Error);

OpmLog::addBackend("LOGFILE" , logFile);
OpmLog::addBackend("CERR" , cerrLog);   
```

we will get both - cool ehh?

**Comment by atgeirr** [comment]
About a fixed hierarchy of message types: I think that would be defined once and for all by the elements of the message type enum. In other contexts I have seen logging levels: each message exists on a level, and which level goes where is up to user and program configuration (possibly at runtime). That is why I expect more severe warnings to appear in every backend that deal with less severe warnings (so Errors may appear in the log file and terminal, while Warnings only go to the log file).

Another thing that struck me: I realize that this is not quite the same as verbosity control either: in sim_fibo_ad now there is some output printed showing the non-linear iterations. Should this be a part of the log system? Probably not, but in that case, how do we control verbosity then (for example only summarizing the number of iterations for each step instead)?

**Comment by atgeirr** [comment]
Another question: the strings you pass when adding a backend (such as CERR or COUNTER): are they just tags that can be completely arbitrary? What would they be used for?

**Comment by akva2** [comment]
just my .02c;

the most convenient loggers i have worked with is a hybrid.

while the storage scheme isn't the point;

N (typically, 3) lower bits are a verbosity level, where higher levels include the lower ones.
some bits are then reserved to message type, like error, warning, etc.

optionally the rest of the bits are then used to allocate "advanced" log components, like for printing residuals and such (or, in my use case; whether to log from specific components). then you can set the verbosity level, the type of the message and which component is logging using simple or's. the component can also be used to split things to several files for instance if that is something one want.

code could then look something like
`my_component = Logger::Get().AllocateComponent("My fancy component", std::cout);
 Logger::Get().Log(INFO|VERY_VERBOSE|my_component, "foobar");`

with the implicit assumption that if no component is specified it is always logged (according to type and verbosity prefs from the user). i have assumed a singleton logging class obviously.

it's very nice from a users pov to be able to say 'high verbosity on warnings, low verbosity on general info'.

**Comment by joakim-hove** [comment]
>Another question: the strings you pass when adding a backend (such as CERR or COUNTER): are they just tags that can be completely arbitrary? What would they be used for?

OK - that one was simple: The `Logger` class has a std::map< std::string,LogBackend >` container to hold on to the backends. The strings 'CERR' and 'COUNTER' can then be used to get hold of the backend reference:

```
auto logBackend = OpmLog::getBackend<StreamLog>("LOGFILE");
```

**Comment by joakim-hove** [comment]
> I realize that this is not quite the same as verbosity control either:

Hmmm - this; and in particular @akva2's later comment made me realize that there might be more to do here ... I will come back to it.

**Comment by joakim-hove** [comment]
> [...] In other contexts I have seen logging levels: each message exists on a level, and which level goes where is up to user and program configuration (possibly at runtime). 

I think we basically agree on this(?). The use of a bitmask has been a convenient choice to facilitate flexible external configuration of which messages go where.

>That is why I expect more severe warnings to appear in every backend that deal with less severe warnings (so Errors may appear in the log file and terminal, while Warnings only go to the log file).

I agree that would be a natural thing to expect; and the user - i.e. developer of `sim_fibo_ad` in this case, should configure which messagetypes go where when configuring the logging backends. So for the case you mention two logging backends would be configured:

```C++
// File supported backend taking both warnings and errors
std::shared_pointer<StreamLog> logFile = std::make_shared<StreamLog>("LOGFILE.TXT" , Log::MessageType::Warning | Log::MessageType::Error);

// Backend writing directly to std::cerr - taking only errors:
std::shared_pointer<StreamLog> cerrLog = std::make_shared<StreamLog>(std::cerr ,  Log::MessageType::Error);

// Add the backends
OpmLog::addBackend("LOGFILE" , logFile);
OpmLog::addBackend("CERR" , cerrLog);


// This message should go both to LOGFILE.TXT and stderr:
OpmLog::addError("This is an error");

// Only to the logfile
OpmLog::addWarning("This is a warning");

// Completely ignored by the two current backends:
OpmLog::addInfo("Some information - ignored");
```

**Comment by joakim-hove** [comment]
>just my .02c;

Thank you for the input - your 2$ are highly appreciated, although I must admit significant parts of it were beyond me.

>N (typically, 3) lower bits are a verbosity level, where higher levels include the lower ones.
some bits are then reserved to message type, like error, warning, etc. [...]  then you can set the verbosity level, the type of the message and which component is logging using simple or's.

In my head the messagetype, i.e. Warning, Error, Bug, ... were the only characteristics of a message, but the way I read you here (and possibly also @atgeirr) you see a log message as having two orthogonal attributes; i.e. both a message type and a verbosity?

**Comment by atgeirr** [comment]
> OK - that one was simple: The Logger class has a std::map< std::string,LogBackend > container to hold on to the backends. The strings 'CERR' and 'COUNTER' can then be used to get hold of the backend reference:

Still, why would one need or want this? I'm not rejecting it, but I'd like to see at least one situation where this would be good to have...

**Comment by akva2** [comment]
yeah, that was the gist of it. with type, verbosity and component it is highly configurable from the user side. with just type you often often end up with having to wade through tons of logging from "irrelevant" parts of the code and SNR decreases.

obviously the verbosity field isn't as relevant for all types but for info, warning and debug in particular they are very handy.

i assume it was the storage scheme that was confusing. it's basically just chopping up an integer into
an int3 for type, int2/int3 for verbosity and an int26 or so for the components. simply a compact way of representing it. you then use |, >>, <<, & to check each part of the bitfield. c++11 also has some fancy bitfield classes iirc which might be more convenient than the oldschool way.

**Comment by atgeirr** [comment]
> i assume it was the storage scheme that was confusing. it's basically just chopping up an integer into
> an int3 for type, int2/int3 for verbosity and an int26 or so for the components. simply a compact way of representing it. you then use |, >>, <<, & to check each part of the bitfield. c++11 also has some fancy bitfield classes iirc which might be more convenient than the oldschool way.

The storage scheme is just an implementation detail, but it does hint to the most important thing: the concepts you try to model. In your case, would you see the types being orthogonal among themselves? I.e. Warning, Note and Error being fully independent notions? Or with a relationship: Note < Warning < Error.

**Comment by akva2** [comment]
independent. you explicitly turn a type on or off (and set the verbosity for the type in the same process).

**Comment by joakim-hove** [comment]
> i assume it was the storage scheme that was confusing.

Well - the storage system had the potential for confusion yes; but my challenge is getting around the mixture of both message types and verbosity. The way I interpret your suggestion, please correct me if I am wrong, is (for clarity I am not using bitmasks of any sort here):

```C++
// Logging implementation:
OpmLog::addInfo( int messageVerbosity , const std::string& message) {
    if (messageVerbosity >= verbosity)
       logMessage( message );
}


// Application code 
// Message which will only be added at verbosity level >= 3
OpmLog::addInfo( 3 , "This is a really useless message");

// Quite important message, coming at verbosity level >= 1:
OpmLog::addInfo( 1 , "Important message");
```

This implies that a verbosity level must be attached to every message in the application code? For Info/Debug this probably makes sense, but for instance Error and Warning should "always" be on a low verbosity I guess?

**Comment by akva2** [comment]
basically, yes. if the dev does not explicitly specify a verbosity the verbosity is 0 - aka always print if logs of that type is enabled. if the dev choose to assign it a higher verbosity level, the user will have to bump the level to get it printed.

**Comment by akva2** [comment]
i.e. you would change the Opm::addInfo() to have a signature like
'Opm::addInfo(const std::string& message, int messageVerbosity=0)`

this obviously doesn't play nice with varargs, which is why it's convenient to encode all info in a single int (not sure if you're gonna use varargs/ellipses or not).

**Comment by joakim-hove** [comment]
>Still, why would one need or want this? I'm not rejecting it, but I'd like to see at least one situation where this would be good to have...

Well - the concrete reason in this case is to be able to query the `CounterLog` instance how many Warnings, Errors, ... there have been. In addition this facilitates the possibility to drop (and maybe modify??) logging backends runtime. But I guess the logging backends could just be in a "opaque" container which does not support any form of random access.

**Comment by joakim-hove** [comment]
>this obviously doesn't play nice with varargs,

I had not intended to use varargs - I thought that was frowned upon in the &uuml;ber-typesafe C++ language?

**Comment by joakim-hove** [comment]
@atgeirr : I think I can make sense of @akva2 suggestions now; would this be an OK solution to you?

**Comment by atgeirr** [comment]
Yes, I think that is a good concept. However I would like the severity levels (Info, Warning, Error) to be escalating, so that a warning log always includes the errors as well etc. I would also like for the implementation details (bitsets or not) not to leak into the interface, so that dev-users would not be expected to use bit-logic to get things right when adding a message (this goes hand in hand with escalating severity levels). I think an integer verbosity level is fine, it seems logical that higher levels are for less important messages.

**Comment by joakim-hove** [comment]
Merging this now in agreement with @atgeirr. There is certainly more to with the logging framework itself, in particular *verbosity*, but the changes in downstream modules `opm-core` and `opm-autodiff` are very small anyway.

**Pull request merged**